ENH: OOB Permutation Importance for Random Forests

robert-robison opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->
This implements out-of-bag permutation importance in random forests. This is why it provides value in addition to `inspection.permutation_importance`:

* It's standard in both main R Random Forest packages: `randomForest` and `Ranger`.
* OOB permutation importances do not require a significant change to the workflow (creating an additional validation set)
* Although it's covered in the documentation, many users likely do not realize just how biased impurity-based metrics are (see [here](https://scikit-learn.org/stable/auto_examples/inspection/plot_permutation_importance.html#sphx-glr-auto-examples-inspection-plot-permutation-importance-py); it marks a random feature as most important).
  * For HistGradientBoostingClassifier/Regressor, impurity-based feature importances were left out entirely for this reason (see [#15132 (comment)](https://github.com/scikit-learn/scikit-learn/issues/15132#issuecomment-581253386))
* OOB permutation importances are faster and more stable than `inspection.permutation_importance` across repeated runs on the same data in the tests I've run (see [here](https://github.com/robert-robison/scikit-learn/blob/perm_imp_benchmarking/permutation_importance_benchmarking.ipynb))
* Would more seamlessly work with `SelectFromModel` (see discussion in #9606)

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
There are no open issues this directly addresses or closes to my knowledge. This has been discussed in multiple issues and PRs over the past several years, some of which are referenced above.

Issues/PRs that have suggested/implemented some type of RF permutation importance have been closed based on `eli5` permutation importance or, more recently, `sklearn.inspection.permutation_importance`:

* #3436
* #8027
* #8898
* #6024
* #16023, where @ogrisel states:

> As we have a model agnostic solution to assert feature importances without suffering from cardinality bias, I would be in favor of not extending our Random Forest implementation just to tackle bias issues in their feature importances, unless the method described in this paper is significantly better than permutation based feature importances but this would need to be demonstrated on some example.

I believe this is demonstrated by the reasons stated above, namely:

* Minimizes misinterpretation due to impurity-based importance
* Simplifies workflow by removing need for additional validation set
* Is more stable and faster

#### What does this implement/fix? Explain your changes.

The specific method that is implemented is the one shown in [Breimann (2001)](https://link.springer.com/content/pdf/10.1023/A:1010933404324.pdf), which is described in more details by the developers of the `Ranger` R package in the paper [Do Little Interactions get Lost in Dark Random Forests?](https://link.springer.com/article/10.1186/s12859-016-0995-8) The algorithm works like this:

* Get the OOB accuracy for each tree.
* Randomly shuffle a feature.
* Get the OOB accuracy again. The average drop in accuracy across all trees is the permutation importance for that feature.
* Repeat for all features.

For regression, the `r2_score` is used instead of accuracy.

This was implemented by adding an `importance_type` parameter that can take the values `{"impurity", "permutation"}`. While impurity-based importances aren't calculated until you need them, you need the dataset to calculate permutation importances, so they are calculated during the fit process if specified.

#### Any other comments?

The changes made here are functionally equivalent to the ones made in #3436, so I did reference that PR often while making these changes.

Also, this is my first pull request and this is kind of a big change, so I completely understand if I'm overstepping bounds here. Thank you!

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

robert-robison committed:
implemented perm imp in oob_score for class and reg

robert-robison committed:
Fixed RandomTreesEmbedding compatability error

robert-robison committed:
Put permutation importance in its own method

robert-robison committed:
added tests and fixed random_state and formatting

robert-robison commented:
Anyone have any thoughts on this?

glemaitre commented:
I think this would be a good addition. However, the feature should be shared for:

- RandomForest
- RandomTreeEmbedding
- ExtraTreees

I think that we should take into consideration the following comment: https://github.com/scikit-learn/scikit-learn/pull/3436#issuecomment-219964436

Using a scorer should be possible: having a private function computing the importance given a scorer. The scorer would be specific and given by the `ForestRegressor` (using r2_score) and `ForestClassifier` (using the accuracy).

In addition, we should add documentation. I think that we should show update the example showing the difference between random forest feature importance and permutation importance. I quickly check and it is true that we can show that the importance of the random feature will be really low.

Regarding the parameter, I think that I would be explicit by making it contain `feature_importances_*` (e.g. `feature_importances_type`). In the future, we can see if we can change the default permutation then.

glemaitre commented on sklearn/ensemble/_forest.py:
```diff
@@ -505,7 +522,72 @@ def __init__(self,
+            max_samples=max_samples,
+            importance_type=importance_type)
+
+    def _set_oob_permutation_importance(self, X, y):
```
So I assume that this function could be moved into the `BaseForest` and could accept a `scorer` parameter.

glemaitre commented on sklearn/ensemble/_forest.py:
```diff
@@ -193,7 +193,8 @@ def __init__(self,
                  class_weight=None,
-                 max_samples=None):
+                 max_samples=None,
+                 importance_type='impurity'):
```
We need to have under `self.importance_type = importance_type` (or whatever named. It means that we will share the parameter between RandomForest, RandomTreesEmbedding, and ExtraTrees.

robert-robison replied:
Added

robert-robison committed:
Add scorer, update example, changed param name

robert-robison committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into rf_permutation_importance

robert-robison committed:
formatting

robert-robison commented:
@glemaitre Thanks for reviewing. I moved `_set_oob_permutation_importance` to `BaseForest` and added a `scoring` parameter to all subclasses for use with `check_scoring`. That way it will default to using the `DecisionTree*.score()` method if nothing is explicitly passed. 

I don't think it makes sense to implement this for `RandomTreeEmbedding` because it's unsupervised. Correct me if I'm wrong.

I also added to the example comparing `inspection.permutation_importance` to default Random Forest feature importance, and changed the parameter name to `feature_importances_type` like you suggested. Let me know your thoughts

glemaitre commented:
The tests are failing. I think that there is something wrong with the example that you modify (`name_steps` instead of `named_steps`)?

robert-robison committed:
fixed bug in example

robert-robison commented:
An extra `named_steps` got in there somehow, now fixed.

glemaitre commented on sklearn/ensemble/_forest.py:
```diff
@@ -398,13 +408,63 @@ def fit(self, X, y, sample_weight=None):
+            )
+
+            for col in range(X.shape[1]):
+                X_permuted = X.copy()
```
And I am almost sure that we could benefit from the `sklearn.inspection._permutation_importance._calculate_permutation_scores` function that given all input will compute the score.


We can also parallelize one of the `for` loop. I would imagine that it would be best to parallelize the external loop, meaning on the number of estimators.

In the other permutation importance scheme, we do multiple rounds. I assume that here we don't need to do that since we rely on the bag of estimators to have an estimate of the variance as well.

glemaitre commented on sklearn/ensemble/_forest.py:
```diff
@@ -398,13 +408,63 @@ def fit(self, X, y, sample_weight=None):
+
+            n_predictions[unsampled_indices] += 1
+
+        self._oob_permutation_importance = all_imp.mean(axis=0)
```
I could imagine that we could store a `Bunch` object with the mean, std and all importance as in `permutation_importance`:

```
    result: Bunch
        Dictionary-like object, with the following attributes.
        importances_mean: ndarray, shape (n_features, )
            Mean of feature importance over n_repeats.
        importances_std: ndarray, shape (n_features, )
            Standard deviation over n_repeats.
        importances: ndarray, shape (n_features, n_repeats)
            Raw permutation importance scores.
```

glemaitre commented on sklearn/ensemble/_forest.py:
```diff
@@ -493,7 +556,9 @@ def __init__(self,
-                 max_samples=None):
+                 max_samples=None,
+                 feature_importances_type='impurity',
+                 scoring=None):
```
I don't think that you should expose the parameter here but instead fix it:

```python
class ForestClassifier(BaseForest, ...):
    def __init__(self, ...):
        super().__init__(..., oob_scoring="accuracy")

glemaitre commented on sklearn/ensemble/_forest.py:
```diff
@@ -747,7 +814,9 @@ def __init__(self,
-                 max_samples=None):
+                 max_samples=None,
+                 feature_importances_type='impurity',
+                 scoring=None):
```
same here

glemaitre commented on sklearn/ensemble/_forest.py:
```diff
@@ -1069,6 +1140,23 @@ class RandomForestClassifier(ForestClassifier):
+          The importance corresponds with the average decrease in r2_score
+          across all tree when a feature is permuted or shuffled.
+
+    scoring : string, callable, or None, default=None
```
I would not expose this as mentioned above. It would only be a non-public attribute for the moment.

glemaitre commented on sklearn/ensemble/_forest.py:
```diff
@@ -1069,6 +1140,23 @@ class RandomForestClassifier(ForestClassifier):
         .. versionadded:: 0.22
 
+    feature_importances_type : {"impurity", "permutation"}, default="impurity"
+        The type of feature importance to calculate:
```
```suggestion
        The type of feature importance to compute:
```

glemaitre commented on sklearn/ensemble/_forest.py:
```diff
@@ -193,7 +193,9 @@ def __init__(self,
-                 max_samples=None):
+                 max_samples=None,
+                 feature_importances_type='impurity',
+                 scoring=None):
```
Since it should not be a public attribute we can call it `self._oob_scoring` with a leading underscore.

glemaitre replied:
Actually we don't need any `scoring` anywhere because `check_scoring(..., scoring=None) will rely on the default score of the estimator so we will be completely fine. You can remove this parameter everywhere.

glemaitre commented on sklearn/ensemble/_forest.py:
```diff
@@ -1069,6 +1140,23 @@ class RandomForestClassifier(ForestClassifier):
+          total reduction of the criterion brought by that feature.  It is also
+          known as the Gini importance.
+        - If "permutation", then gets out-of-bag permutation importance.
+          The importance corresponds with the average decrease in r2_score
```
```suggestion
          The importance corresponds with the average decrease in R2
```

glemaitre replied:
Actually, it should be the accuracy for classifier

glemaitre commented on sklearn/ensemble/_forest.py:
```diff
@@ -1069,6 +1140,23 @@ class RandomForestClassifier(ForestClassifier):
+          known as the Gini importance.
+        - If "permutation", then gets out-of-bag permutation importance.
+          The importance corresponds with the average decrease in r2_score
+          across all tree when a feature is permuted or shuffled.
```
```suggestion
          across all trees when a feature is permuted or shuffled.
          
       .. versionadded: 1.0
```

glemaitre commented on sklearn/ensemble/_forest.py:
```diff
@@ -1366,6 +1457,23 @@ class RandomForestRegressor(ForestRegressor):
+          known as the Gini importance.
+        - If "permutation", then gets out-of-bag permutation importance.
+          The importance corresponds with the average decrease in r2_score
+          across all tree when a feature is permuted or shuffled.
```
Same remark here.

glemaitre commented on sklearn/ensemble/_forest.py:
```diff
@@ -1679,6 +1790,23 @@ class ExtraTreesClassifier(ForestClassifier):
+          known as the Gini importance.
+        - If "permutation", then gets out-of-bag permutation importance.
+          The importance corresponds with the average decrease in r2_score
+          across all tree when a feature is permuted or shuffled.
```
same remark here

glemaitre commented on sklearn/ensemble/_forest.py:
```diff
@@ -2290,6 +2441,10 @@ def __init__(self,
         self.sparse_output = sparse_output
 
+    def _set_oob_permutation_importance(self, X, y):
+        raise NotImplementedError("OOB permutation importance not supported "
```
Since we don't expose the parameter to compute the feature importance via permutation, we don't need this safeguard.

glemaitre commented on sklearn/ensemble/tests/test_forest.py:
```diff
@@ -371,6 +376,86 @@ def test_unfitted_feature_importances(name):
+
+        assert_raises(
+            NotImplementedError,
+            ForestEstimator()._set_oob_permutation_importance,
```
This is a private function so no need to test.

glemaitre commented on sklearn/ensemble/tests/test_forest.py:
```diff
@@ -371,6 +376,86 @@ def test_unfitted_feature_importances(name):
+def test_oob_importances_raise_error(name):
+    ForestEstimator = FOREST_ESTIMATORS[name]
+
+    if name in FOREST_TRANSFORMERS:
```
Here, you should be able to parametrize the function regarding the type of error raised, the error message and the parameter passed to the estimator and just call `estimator.fit(X, y)` to check that the error is raised.

Use pytest to check the error message

```python
with pytest.raises(ValueError, match="partial error message"):
    call_raising_error(....)
```

glemaitre commented on sklearn/ensemble/tests/test_forest.py:
```diff
@@ -371,6 +376,86 @@ def test_unfitted_feature_importances(name):
+
+
+@pytest.mark.parametrize("name", FOREST_CLASSIFIERS_REGRESSORS)
+@pytest.mark.parametrize("dtype", (np.float64, np.float32))
```
Do we actually need to parametrize regarding the dtype?

glemaitre commented on sklearn/ensemble/tests/test_forest.py:
```diff
@@ -371,6 +376,86 @@ def test_unfitted_feature_importances(name):
+@pytest.mark.parametrize("name", FOREST_CLASSIFIERS_REGRESSORS)
+@pytest.mark.parametrize("dtype", (np.float64, np.float32))
+def test_oob_importances_multi_class(name, dtype):
+
```
Put a comment to explain what you are testing and why the assertion should be this way.

glemaitre commented on sklearn/ensemble/tests/test_forest.py:
```diff
@@ -371,6 +376,86 @@ def test_unfitted_feature_importances(name):
 
 
+@pytest.mark.parametrize("name", FOREST_CLASSIFIERS_REGRESSORS)
+@pytest.mark.parametrize("dtype", (np.float64, np.float32))
```
same here

glemaitre reviewed (commented):
I would add a test to check that the permutation implemented does not suffer from the bias as the `feature_importances_` using impurity. You can create a random variable and check that it is important with the impurity but not important with the permutation technique.

glemaitre commented on sklearn/ensemble/_forest.py:
```diff
@@ -398,13 +408,63 @@ def fit(self, X, y, sample_weight=None):
+            for col in range(X.shape[1]):
+                X_permuted = X.copy()
+                random_state.shuffle(X_permuted[:, col])
+                curr_perf = scorer(
```
Actually, I am having a second thought here. I give an additional read to Breiman's paper after looking at our function that computes the OOB score.

In your proposal, we average the score of each individual tree while I think that the idea would be to aggregate the votes of each tree and finally (I assume with majority vote / mean) compare with the true target. This is indeed what is intended in the OOB scoring function.

I am not sure that both approaches are equivalent. In this way it seems that we should have a function that should build the OOB predictions that could be reuse between the OOB score and the permutation OOB.

@NicolasHug @ogrisel Could you help regarding the theoretical part.

Then I assume that we can fusion the OOB scoring where we can compute the permutation score if required instead of doing the same manipulation twice.

robert-robison replied:
I considered that approach when writing this PR, and it might be what Breiman is suggesting, although he doesn't go into much detail. To be clear, these are the two options:

1. Get OOB performance degradation for each tree individually when a feature is permuted, then average those degradations across all trees (proposed in this PR)
2. Get OOB performance degradation across the whole RF when a feature is permuted

I agree with you that the 2nd sounded better to me at first, but I chose the first for 2 reasons:
* It was much more stable in the tests I ran (i.e. importances didn't change as much due to randomness)
* It's the same process used in the R `Ranger` package, as detailed [here](https://link.springer.com/article/10.1186/s12859-016-0995-8):
> The basic idea of the permutation variable importance approach [18] is to consider a variable important if it has a positive effect on the prediction performance. To evaluate this, a tree is grown in the first step, and the prediction accuracy in the OOB observations is calculated. In the second step, any association between the variable of interest x i and the outcome is broken by permuting the values of all individuals for x i , and the prediction accuracy is computed again. The difference between the two accuracy values is the permutation importance for x i from a single tree. The average of all tree importance values in a random forest then gives the random forest permutation importance of this variable. The procedure is repeated for all variables of interest. The package ranger [24] was used in our analyses.

I'm happy to change it to the second option if that's preferred. My vote would be for the first though

robert-robison committed:
Apply suggestions from code review

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

robert-robison committed:
parallelized, removed scoring, fixed tests

robert-robison committed:
formatting

robert-robison committed:
Add random feature test

robert-robison committed:
formatting

robert-robison commented:
I attempted to address all comments. I realized we're basically replicating `inspection.permutation_importance` with a Decision Tree as the estimator on the out-of-bag data, so I changed it to call `permutation_importance` on every estimator in the RF in the new `_get_tree_oob_performance` method. I ran into a circular import error, though, so the import is temporarily in that method. Not sure the best way around that. I also attempted to make the outer for loop parallel; not sure I did that correctly. 

`inspection.permutation_importance` doesn't allow sparse, so I left `sparse=False` for the moment.

All your other suggestions were pretty straightforward. Let me know if I missed anything or messed anything up.

glemaitre commented:
I will review it soon. With @ogrisel, we checked more into details the paper and indeed the phrasing in the paper of Breiman is quite confusing and I think that the current implementation is in line with the one stated in different paper from the literature and the Fortran implementation of Breiman and Cutler (https://math.usu.edu/~adele/forests/cc_home.htm#varimp)

> I attempted to address all comments. I realized we're basically replicating inspection.permutation_importance with a Decision Tree as the estimator on the out-of-bag data, so I changed it to call permutation_importance on every estimator in the RF in the new _get_tree_oob_performance method. I ran into a circular import error, though, so the import is temporarily in that method. Not sure the best way around that. I also attempted to make the outer for loop parallel; not sure I did that correctly.

I have to check as well more into details but it was my comment regarding using `_compute_permutation_importance`. I think that we should use this function instead of `permutation_importance` if possible because it should avoid a couple of useless check (but I have to check).

In the meanwhile, I have been working on https://github.com/scikit-learn/scikit-learn/pull/19162 to improve the way the OOB score is computed and I will probably have a new look to see if it allows for a better modularity with your code (using `_get_oob_predictions` for instance).

glemaitre commented:
The refactoring in https://github.com/scikit-learn/scikit-learn/pull/19162 has been merged. @robert-robison you can resolve the conflict and check if you can benefit from it. Then, I will review your proposal againn.

robert-robison committed:
Merge remote-tracking branch 'upstream/main' into rf_permutation_importance

# Conflicts:
#	sklearn/ensemble/tests/test_forest.py

robert-robison committed:
remove inspection dependency

robert-robison commented:
@glemaitre Conflict resolved; ready for review. I did not use any of the changes because I think using `estimator.score(X[unsampled_indices, :], y[unsampled_indices])` is a little cleaner, but I'm happy to change and incorporate if you'd prefer.

glemaitre commented:
> I did not use any of the changes because I think using estimator.score(X[unsampled_indices, :], y[unsampled_indices]) is a little cleaner.

I think the trick here is to use a `scorer` with `check_scoring(estimator, scorer=None)`. But I will propose another refactoring using the idea that your earlier mentioned using the `permutation_importance`.

glemaitre commented:
In short, I would like to integrate your code directly in `_compute_oob_predictions` and set the importance in `_set_oob_score_and_attributes` and leverage the use of `sklearn.inspection.permutation_importance`.

As you mentioned before, we can use directly either `permutation_importance` of `_calculate_permutation_scores`. The second function would be slightly more efficient because we could only compute the baseline once by hand using the `y_pred` in the `_compute_oob_predictions`. However, it makes the code a bit more complex where we would need a scorer and a score function (the baseline would need `score(y_true, y_pred)`, while the function requires a proper scorer`) and for the sake of maintainability, I would use `permutation_importance`. 

I would introduce the change in `compute_oob_predictions` (we will probably have to rename it then `_compute_oob_predictions_and_importances`) because the way to loop over the estimator and get the OOB samples is the same. So once, the OOB samples are obtained and the predictions accumulated (to compute the OOB score later on), we can compute the permutation importance. We only need a single repeat and therefore, we can store the permutation of all trees in an array `oob_permutation_importances` that would be returned by the function. If not permutation is required, we return `None`.

Finally, we need `_set_oob_score_and_attributes` to set the `self._oob_permutation_importance` if needed by creating a Bunch with all permutation, the mean, and std.

I attached a rough diff to give more details regarding where to introduce the changes.
WDYT? Is there any reason that you went back from using `permutation_importance`?

<details>

```diff
diff --git a/sklearn/ensemble/_forest.py b/sklearn/ensemble/_forest.py
index 0dd6ec5cb1..67027a2217 100644
--- a/sklearn/ensemble/_forest.py
+++ b/sklearn/ensemble/_forest.py
@@ -52,7 +52,7 @@ from joblib import Parallel
 
 from ..base import is_classifier
 from ..base import ClassifierMixin, RegressorMixin, MultiOutputMixin
-from ..metrics import accuracy_score, r2_score
+from ..metrics import accuracy_score, check_scoring, r2_score
 from ..preprocessing import OneHotEncoder
 from ..tree import (DecisionTreeClassifier, DecisionTreeRegressor,
                     ExtraTreeClassifier, ExtraTreeRegressor)
@@ -419,7 +419,7 @@ class BaseForest(MultiOutputMixin, BaseEnsemble, metaclass=ABCMeta):
                     f"supported: continuous, continuous-multioutput, binary, "
                     f"multiclass, multilabel-indicator."
                 )
-            self._set_oob_score_and_attributes(X, y)
+            self._set_oob_score_and_attributes(X, y, sample_weight)
 
         if self.feature_importances_type == "permutation":
             self._set_oob_permutation_importance(X, y, sample_weight)
@@ -500,7 +500,7 @@ class BaseForest(MultiOutputMixin, BaseEnsemble, metaclass=ABCMeta):
             The target matrix.
         """
 
-    def _compute_oob_predictions(self, X, y):
+    def _compute_oob_predictions(self, X, y, sample_weight):
         """Compute and set the OOB score.
 
         Parameters
@@ -509,6 +509,8 @@ class BaseForest(MultiOutputMixin, BaseEnsemble, metaclass=ABCMeta):
             The data matrix.
         y : ndarray of shape (n_samples, n_outputs)
             The target matrix.
+        sample_weight : ndarray of shape (n_samples,)
+            Sample weights used in permutation importances.
 
         Returns
         -------
@@ -517,8 +519,9 @@ class BaseForest(MultiOutputMixin, BaseEnsemble, metaclass=ABCMeta):
             The OOB predictions.
       """
         X = check_array(X, dtype=DTYPE, accept_sparse='csr')
+        random_state = check_random_state(self.random_state)
 
-        n_samples = y.shape[0]
+        n_samples, n_features = X.shape
         n_outputs = self.n_outputs_
         if is_classifier(self) and hasattr(self, "n_classes_"):
             # n_classes_ is a ndarray at this stage
@@ -531,23 +534,43 @@ class BaseForest(MultiOutputMixin, BaseEnsemble, metaclass=ABCMeta):
             # the array operations compatible with the 2 settings
             oob_pred_shape = (n_samples, 1, n_outputs)
 
+        if self.feature_importances_type == "permutation":
+            oob_importances = np.zeros(shape=(n_features, self.n_estimators))
+        else:
+            oob_importances = None
+
         oob_pred = np.zeros(shape=oob_pred_shape, dtype=np.float64)
         n_oob_pred = np.zeros((n_samples, n_outputs), dtype=np.int64)
 
         n_samples_bootstrap = _get_n_samples_bootstrap(
             n_samples, self.max_samples,
         )
-        for estimator in self.estimators_:
+        for idx, estimator in enumerate(self.estimators_):
             unsampled_indices = _generate_unsampled_indices(
                 estimator.random_state, n_samples, n_samples_bootstrap,
             )
+            X_oob, y_oob = X[unsampled_indices, :], y[unsampled_indices]
 
-            y_pred = self._get_oob_predictions(
-                estimator, X[unsampled_indices, :]
-            )
-            oob_pred[unsampled_indices, ...] += y_pred
+            y_oob_pred = self._get_oob_predictions(estimator, X_oob)
+            oob_pred[unsampled_indices, ...] += y_oob_pred
             n_oob_pred[unsampled_indices, :] += 1
 
+            if self.feature_importances_type == "permutation":
+                # avoid circular dependence
+                from ..inspection import permutation_importance
+
+                result_importances = permutation_importance(
+                    estimator,
+                    X_oob,
+                    y_oob,
+                    scoring=None,
+                    n_repeats=1,
+                    n_jobs=self.n_jobs,
+                    random_state=random_state,
+                    sample_weight=sample_weight,
+                )
+                oob_importances[:, idx] = result_importances.importances[:, 0]
+
         for k in range(n_outputs):
             if (n_oob_pred == 0).any():
                 warn(
@@ -558,7 +581,7 @@ class BaseForest(MultiOutputMixin, BaseEnsemble, metaclass=ABCMeta):
                 n_oob_pred[n_oob_pred == 0] = 1
             oob_pred[..., k] /= n_oob_pred[..., [k]]
 
-        return oob_pred
+        return oob_pred, oob_importances
 
     def _validate_y_class_weight(self, y):
         # Default implementation
@@ -690,7 +713,15 @@ class ForestClassifier(ClassifierMixin, BaseForest, metaclass=ABCMeta):
             y_pred = np.rollaxis(y_pred, axis=0, start=3)
         return y_pred
 
-    def _set_oob_score_and_attributes(self, X, y):
+    @staticmethod
+    def _reshape_oob_predictions(oob_predictions):
+        """Reshape OOB predictions specifically for classifier."""
+        if oob_predictions.shape[-1] == 1:
+            # drop the n_outputs axis if there is a single output
+            oob_predictions = oob_predictions.squeeze(axis=-1)
+        return oob_predictions
+
+    def _set_oob_score_and_attributes(self, X, y, sample_weight):
         """Compute and set the OOB score and attributes.
 
         Parameters
@@ -699,15 +730,15 @@ class ForestClassifier(ClassifierMixin, BaseForest, metaclass=ABCMeta):
             The data matrix.
         y : ndarray of shape (n_samples, n_outputs)
             The target matrix.
+        sample_weight : ndarray of sahpe (n_samples, n_outputs)
+            Sample weights applied during scoring.
         """
-        self.oob_decision_function_ = super()._compute_oob_predictions(X, y)
-        if self.oob_decision_function_.shape[-1] == 1:
-            # drop the n_outputs axis if there is a single output
-            self.oob_decision_function_ = self.oob_decision_function_.squeeze(
-                axis=-1
-            )
+        self.oob_decision_function_ = self._reshape_oob_predictions(
+            super()._compute_oob_predictions(X, y, sample_weight)
+        )
         self.oob_score_ = accuracy_score(
-            y, np.argmax(self.oob_decision_function_, axis=1)
+            y, np.argmax(self.oob_decision_function_, axis=1),
+            sample_weight=sample_weight,
         )
 
     def _validate_y_class_weight(self, y):
@@ -983,7 +1014,16 @@ class ForestRegressor(RegressorMixin, BaseForest, metaclass=ABCMeta):
             y_pred = y_pred[:, np.newaxis, :]
         return y_pred
 
-    def _set_oob_score_and_attributes(self, X, y):
+    @staticmethod
+    def _reshape_oob_predictions(oob_predictions):
+        """Reshape OOB predictions specifically for regressor."""
+        oob_predictions = oob_predictions.squeeze(axis=1)
+        if oob_predictions.shape[-1] == 1:
+            # drop the n_outputs axis if there is a single output
+            oob_predictions = oob_predictions.squeeze(axis=-1)
+        return oob_predictions
+
+    def _set_oob_score_and_attributes(self, X, y, sample_weight):
         """Compute and set the OOB score and attributes.
 
         Parameters
@@ -992,14 +1032,15 @@ class ForestRegressor(RegressorMixin, BaseForest, metaclass=ABCMeta):
             The data matrix.
         y : ndarray of shape (n_samples, n_outputs)
             The target matrix.
+        sample_weight : ndarray of shape (n_samples,)
+            Sample weights used for scoring.
         """
-        self.oob_prediction_ = super()._compute_oob_predictions(X, y).squeeze(
-            axis=1
+        self.oob_prediction_ = self._reshape_oob_predictions(
+            super()._compute_oob_predictions(X, y, sample_weight)
+        )
+        self.oob_score_ = r2_score(
+            y, self.oob_prediction_, sample_weight=sample_weight
         )
-        if self.oob_prediction_.shape[-1] == 1:
-            # drop the n_outputs axis if there is a single output
-            self.oob_prediction_ = self.oob_prediction_.squeeze(axis=-1)
-        self.oob_score_ = r2_score(y, self.oob_prediction_)
 
     def _compute_partial_dependence_recursion(self, grid, target_features):
         """Fast partial dependence computation.
```

glemaitre commented:
Once that we start to settle on where to integrate the computation, we will need to make a choice regarding the parallelization of the code. For the moment, you parallelize on the outer loop for the tree. The permutation importance also parallelizes on the feature. So we will need to profile a bit a sequential version to see which part of computing the OOB permutation importance is costly and decide which backend to use (i.e. threading or multiprocessing) and which loop to parallelize (or both).

robert-robison commented:
Nice, I like that. I will make the changes in the next couple of days. The only reason I took out `permutation_importance` is because of the circular dependence. If it's okay to have that, then I would also prefer to use it.

robert-robison committed:
integrate permutation imp with oob score

robert-robison commented:
@glemaitre Made the changes, let me know what you think. Seemed a bit slower without the outer loop being parallelized

glemaitre commented:
OK, the implementation is neat. So now, we can start to profile and see which parallelization would be best.
I will experiment and come back with some results from which we can decide what will be the way forward.

glemaitre commented on sklearn/ensemble/_forest.py:
```diff
@@ -915,14 +971,27 @@ def _set_oob_score_and_attributes(self, X, y):
             # drop the n_outputs axis if there is a single output
             self.oob_prediction_ = self.oob_prediction_.squeeze(axis=-1)
-        self.oob_score_ = r2_score(y, self.oob_prediction_)
+        self.oob_score_ = r2_score(
+            y, self.oob_prediction_, sample_weight=sample_weight
+        )
```
```suggestion
        if self.oob_score:
            self.oob_score_ = r2_score(
                y, self.oob_prediction_, sample_weight=sample_weight
            )
```

glemaitre replied:
It also means that we need to add a test to ensure that `oob_score_` is not set when `oob_score=False` even when the permutation importance is computed.

glemaitre commented on sklearn/ensemble/_forest.py:
```diff
@@ -915,14 +971,27 @@ def _set_oob_score_and_attributes(self, X, y):
+            y, self.oob_prediction_, sample_weight=sample_weight
+        )
+
+        if self.feature_importances_type == "permutation":
```
I would prefer the following:

 ```suggestion
         if oob_importances is not None:
```

glemaitre commented on sklearn/ensemble/_forest.py:
```diff
@@ -915,14 +971,27 @@ def _set_oob_score_and_attributes(self, X, y):
         """
-        self.oob_prediction_ = super()._compute_oob_predictions(X, y).squeeze(
-            axis=1
+        self.oob_prediction_, oob_importances_ = (
```
It would be enough to the final underscore when creating the attribute (this is the convention for fitted attribute).

```suggestion
        self.oob_prediction_, oob_importances = (
```

glemaitre reviewed (commented):
Just something that I saw while playing with the code

glemaitre commented on sklearn/ensemble/_forest.py:
```diff
@@ -624,17 +664,31 @@ def _set_oob_score_and_attributes(self, X, y):
+            sample_weight=sample_weight
         )
 
+        if self.feature_importances_type == "permutation":
```
I am proposing the following patch avoiding the duplication for setting the permutation importance in the specialized method:

<details>

```diff
diff --git a/sklearn/ensemble/_forest.py b/sklearn/ensemble/_forest.py
index 7bbfdcf320..6997382859 100644
--- a/sklearn/ensemble/_forest.py
+++ b/sklearn/ensemble/_forest.py
@@ -432,6 +432,11 @@ class BaseForest(MultiOutputMixin, BaseEnsemble, metaclass=ABCMeta):
     def _set_oob_score_and_attributes(self, X, y, sample_weight):
         """Compute and set the OOB score and attributes.
 
+        This abstract method compute the predictions and the importances
+        (if required) and will set the permutation importance attributes.
+        The predictions are returned to the child methods to set the
+        specialized class attributes.
+
         Parameters
         ----------
         X : array-like of shape (n_samples, n_features)
@@ -441,6 +446,19 @@ class BaseForest(MultiOutputMixin, BaseEnsemble, metaclass=ABCMeta):
         sample_weight : ndarray of shape (n_samples,)
             Sample weights.
         """
+        oob_predictions, oob_importances = \
+            self._compute_oob_predictions_and_importances(
+                X, y, sample_weight
+            )
+
+        if oob_importances is not None:
+            self._oob_permutation_importance = Bunch(
+                importances_mean=np.mean(oob_importances, axis=1),
+                importances_std=np.std(oob_importances, axis=1),
+                importances=oob_importances
+            )
+
+        return oob_predictions
 
     def _compute_oob_predictions_and_importances(self, X, y, sample_weight):
         """Compute and set the OOB score.
@@ -667,9 +685,8 @@ class ForestClassifier(ClassifierMixin, BaseForest, metaclass=ABCMeta):
         sample_weight : ndarray of shape (n_samples,)
             Sample weights.
         """
-        self.oob_decision_function_, oob_importances_ = (
-            super()
-            ._compute_oob_predictions_and_importances(X, y, sample_weight)
+        self.oob_decision_function_ = super()._set_oob_score_and_attributes(
+            X, y, sample_weight
         )
 
         if self.oob_decision_function_.shape[-1] == 1:
@@ -677,16 +694,11 @@ class ForestClassifier(ClassifierMixin, BaseForest, metaclass=ABCMeta):
             self.oob_decision_function_ = self.oob_decision_function_.squeeze(
                 axis=-1
             )
-        self.oob_score_ = accuracy_score(
-            y, np.argmax(self.oob_decision_function_, axis=1),
-            sample_weight=sample_weight
-        )
 
-        if self.feature_importances_type == "permutation":
-            self._oob_permutation_importance = Bunch(
-                importances_mean=np.mean(oob_importances_, axis=1),
-                importances_std=np.std(oob_importances_, axis=1),
-                importances=oob_importances_
+        if self.oob_score:
+            self.oob_score_ = accuracy_score(
+                y, np.argmax(self.oob_decision_function_, axis=1),
+                sample_weight=sample_weight
             )
 
     def _validate_y_class_weight(self, y):
@@ -974,23 +986,17 @@ class ForestRegressor(RegressorMixin, BaseForest, metaclass=ABCMeta):
         sample_weight : ndarray of shape (n_samples,)
             Sample weights.
         """
-        self.oob_prediction_, oob_importances_ = (
-            super()
-            ._compute_oob_predictions_and_importances(X, y, sample_weight)
+        self.oob_prediction_ = super()._set_oob_score_and_attributes(
+            X, y, sample_weight
         )
         self.oob_prediction_ = self.oob_prediction_.squeeze(axis=1)
         if self.oob_prediction_.shape[-1] == 1:
             # drop the n_outputs axis if there is a single output
             self.oob_prediction_ = self.oob_prediction_.squeeze(axis=-1)
-        self.oob_score_ = r2_score(
-            y, self.oob_prediction_, sample_weight=sample_weight
-        )
 
-        if self.feature_importances_type == "permutation":
-            self._oob_permutation_importance = Bunch(
-                importances_mean=np.mean(oob_importances_, axis=1),
-                importances_std=np.std(oob_importances_, axis=1),
-                importances=oob_importances_
+        if self.oob_score:
+            self.oob_score_ = r2_score(
+                y, self.oob_prediction_, sample_weight=sample_weight
             )
 
     def _compute_partial_dependence_recursion(self, grid, target_features):
```

glemaitre replied:
In this way, we only specialized the predictions/decision_function and score for each class.

glemaitre committed:
MNT refactoring based on further multiprocessing

glemaitre commented:
I am going to push a couple of commits after profiling the code.

To summarize:

- We can go with multiprocessing on the outer loop. There is a small gain there but I only profiled on a machine with 2 CPUs.
- We should not parallelize the inner loop within the permutation importance. Having `n_repeats=1` is limiting and makes that the overhead of multiprocessing is too large. Even if we release the GIL in the tree prediction, using threading there is not helpful.
- Due to these choices, then we should decouple the OOB scoring and importance computation. Indeed, the OOB score is computed by collecting predictions that relies on an accumulation in a NumPy array. This cannot be used in a shared-memory parallel setting. We would need to store all the predictions from all the trees. It might be really memory costly.
- We should also disable whenever possible the underlying input checks for the tree prediction and scoring.

I pushed a first PR to decouple the processing but I will push a subsequent commit for the multiprocessing and the bypass of the input validation.

glemaitre committed:
doc

glemaitre committed:
ENH parallelize

glemaitre committed:
doc

glemaitre commented:
In addition, after discussing with @ogrisel and @jeremiedbb, I renamed `feature_importances_type` to `feature_importances` (similar to `oob_score` and the fitted attribute `oob_score_`) and `"permutation"` to `"permutation_oob"` to make it explicit that the permutation is done on the OOB samples.

glemaitre committed:
less diff

glemaitre committed:
improve doc

glemaitre committed:
TST check for features importances raised error

glemaitre commented on sklearn/ensemble/tests/test_forest.py:
```diff
@@ -373,6 +378,136 @@ def test_unfitted_feature_importances(name):
+
+@pytest.mark.parametrize("name", FOREST_CLASSIFIERS_REGRESSORS)
+def test_oob_importances_multi_class(name):
+    # Check that oob permutation importances correctly identify that
```
I don't think that we need this test since that the feature importance will not be linked to the number of classes

glemaitre commented on sklearn/ensemble/tests/test_forest.py:
```diff
@@ -373,6 +378,136 @@ def test_unfitted_feature_importances(name):
 
 
+@pytest.mark.parametrize("name", FOREST_CLASSIFIERS_REGRESSORS)
+def test_oob_importances(name):
```
I would separate the classification and regression and have a dedicated set for each case.
I think that you can generate the data in the test itself. It is not too costly to do so.

glemaitre commented on sklearn/ensemble/tests/test_forest.py:
```diff
@@ -373,6 +378,136 @@ def test_unfitted_feature_importances(name):
+    # important using impurity-based feature importance but not out-of-bag
+    # permutation importance
+
+    X = X_large.copy()
```
Let's separate regression/classification and we can generate the dataset in the function directly.

glemaitre commented on sklearn/ensemble/tests/test_forest.py:
```diff
@@ -373,6 +378,136 @@ def test_unfitted_feature_importances(name):
+    X = X_large.copy()
+    y = y_large.copy()
+
+    X[:, :-1] = (X[:, :-1] > 0).astype(int)
```
A comment to explain what we intend here would be good.

glemaitre commented on sklearn/ensemble/_forest.py:
```diff
@@ -615,24 +735,29 @@ def _get_oob_predictions(tree, X):
         self.oob_score_ = accuracy_score(
-            y, np.argmax(self.oob_decision_function_, axis=1)
+            y, np.argmax(self.oob_decision_function_, axis=1),
+            sample_weight=sample_weight
```
Here, this is a bit straightforward addition regarding `sample_weight` but I don't recall that we have a test for that.

It could be good to add that having the correspondence between `sample_weight=None` and `sample_weight=np.ones(...)`.

It would also be nice to come up with a test where the weights are non-unit ones. It might be a bit tricker but I don't have a good example right now.

glemaitre commented on examples/inspection/plot_permutation_importance.py:
```diff
@@ -172,3 +172,33 @@
 fig.tight_layout()
 plt.show()
+
+# %%
```
IRL with @ogrisel, he was proposing to put this example under the MDI computation.
Additionally, we should mention that MDI and permutation on OOB are only computed on the feature space provided to the forest (i.e. after the preprocessing) and that the last function allows inspecting at different stages.

In addition, we should make it explicit that MDI is computed from the in-sample and would suffer from a bias if the model learnt overfits. The permutation OOB offers an alternative similar to the external permutation importance. It is even better if we have limited data since we are not required to split the data into training/testing and leverage the OOB.

Finally, we should mention that the estimate of the variance of the permutation importance could be more efficient with the OOB scheme: `n_repeats` would be required to be set to `n_estimators` to have as many estimates.

glemaitre reviewed (commented):
In addition to the current comment, we need to update the User Guide: https://scikit-learn.org/dev/modules/ensemble.html#feature-importance-evaluation

I would like to have another deeper look into the test to be sure that we don't miss anything.

glemaitre commented:
I think that we can think to integrate this feature in 1.0. Please add an entry to the change log at `doc/whats_new/v*.rst`. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`.

robert-robison committed:
reformat tests, update example

robert-robison commented:
I think I addressed all your comments regarding tests, the example, and the change log. Let me know if I missed anything. Not sure why the tests are failing

glemaitre commented:
The failure is unrelated. The server from OpenML (to fetch some dataset) is down.

glemaitre committed:
DOC rework the example

glemaitre committed:
glitch

glemaitre committed:
DOC update user guide

glemaitre committed:
DOC solve title marker

glemaitre committed:
DOC improve example regarding feature importance

glemaitre committed:
DOC add new attributes importances_

glemaitre committed:
DOC update whats new

glemaitre committed:
update docstring feature_importances_

glemaitre committed:
TST add test for importances_ attribute

glemaitre committed:
PEP8

glemaitre committed:
TST improve couple of assert

glemaitre committed:
DOC add missing documentation

glemaitre committed:
clean-up

glemaitre committed:
small fix

glemaitre commented:
@robert-robison I mainly pushed changes regarding the documentation. It was easier to make the push than ask for the changes on files that were not modified.

I also added an attribute `importances_` that contains the `Bunch` for both impurity-based and permutation feature importance. In this way, we have the same API. From changes an example, it seems that this is also something that we needed.

glemaitre committed:
DOC use boxplot for all plot in example

glemaitre reviewed (approved):
I think that the PR is in good shape for another review.
Probably @ogrisel can have a look at it since we already discuss the PR.

glemaitre committed:
EXA solve issue cutted ylabel

glemaitre committed:
DOC add support for sample_weight in OOB score

glemaitre committed:
style code

glemaitre committed:
DOC add a note regarding correlated feature

cmarmo commented:
Hi @robert-robison thanks for your work so far! Do you mind synchronizing with upstream? This will help in triggering a second review. Thanks!

robert-robison committed:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into rf_permutation_importance

robert-robison committed:
Attempting to resolve merge in examples

robert-robison committed:
formatting

robert-robison committed:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into rf_permutation_importance

robert-robison committed:
formatting

robert-robison commented:
@cmarmo Just updated, although not sure I resolved all conflicts correctly, especially with `examples/ensemble/plot_forest_importances.py`. Might be helpful to have @glemaitre check my work and fix any errors

ogrisel committed:
Merge branch 'main' into rf_permutation_importance

jjerphan reviewed (changes requested):
Hi @robert-robison, thanks for this PR.

To me, this is a notable contribution. As MDI has shown to be biased in some cases, adding an alternative and documenting it with examples entirely makes sense. :+1: 

I did a first pass on the example and part of the docs. I am planning to review the code of the implementation after interacting on this review. :)

thomasjpfan committed:
MAINT Adds target_version to black config (#20293)

robert-robison commented on examples/inspection/plot_permutation_importance.py:
```diff
@@ -118,57 +152,146 @@
+# importances of these two features, we see that the reported feature
+# importances have a large standard deviation. Indeed, a tree could use either
+# variables to make a split because both variable carry the exact same
+# information. Thus, one tree could use the feature `sex_male` and ignore
+# `sex_female` to create split while another tree could could make the opposite
+# choice. We will see that the permutation feature importance does not solve
+# this issue.
```
No, it was correct before. It does not solve the issue

robert-robison commented on examples/inspection/plot_permutation_importance.py:
```diff
@@ -118,57 +152,146 @@
+# As presented, the feature permutation importances can be computed either
+# on the OOB samples or on separated datasets.
+#
+# While they are similar, it should be noted that the variations of the
+# importances is estimated differently: the variance of the decrease of the
+# score is estimated across the number of trees (i.e. `n_estimators` parameter)
+# in the forest while it is estimated via the number of repeated permutation
+# (i.e. `n_repeats`) in the other strategy.
```
Looks good! I think 'drown' should be 'done', though, right?

robert-robison replied:
Actually after reading this again, I would argue that the original formulation is clearer. The point being made is the _variations_ of the importances are estimated differently between the methods. That's more clear to me the way it currently is.

jjerphan replied:
I see. Probably using variance over variations in the original text would make it clearer? What do you think?

As of "drown", this is a mistake from me. "Made" is probably better in this case.

robert-robison commented on examples/inspection/plot_permutation_importance.py:
```diff
@@ -118,57 +152,146 @@
+# choice. We will see that the permutation feature importance does not solve
+# this issue.
+#
+# Alternative to MDI using Feature Permutation Importance
+# -------------------------------------------------------
+# The limitations of MDI pointed out in the previous section can be bypassed
+# using an alternative strategy to estimate the feature importances. This
+# strategy relies on monitoring the decrease (or not) of a given performance
+# metric by randomly permutting the value of a given feature. In short, a
+# predictive feature will negatively impact the score when it is randomly
+# permuted while a non-predictive feature will not change the score.
+#
+# This feature permutation importance estimate can be computed in two different
+# way: (i) by using the out-of-bag (OOB) samples in the ensemble to perform the
+# permutation and the scoring or (ii) by manually splitting and handling a
+# train and test set where the latter will be used with permutations.
+#
+# Feature Permutation Importance on Out-Of-Bag (OOB) samples
```
Both this and the previous suggestions look great, but they modify the same lines...not sure the way that's supposed to be handled

jjerphan replied:
In this case, the best is to make the modification locally. This also allow grouping suggestions.

robert-robison committed:
Apply suggestions from code review

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

robert-robison commented:
@jjerphan Thanks for the review! Committed most of the suggestions, although some were modifying the same lines, so I wasn't sure how to handle those. Let me know what else I can do to help the process

robert-robison committed:
Apply additional suggestions from code review

jjerphan commented on examples/inspection/plot_permutation_importance.py:
```diff
@@ -118,57 +151,151 @@
+# Alternative to MDI using Permutation Feature Importance
+# -------------------------------------------------------
+# The limitations of MDI pointed out in the previous section can be mitigated
+# using an alternative strategy: Feature Permutation Importance.
```
Small nitpick. :slightly_smiling_face: 

```suggestion
# using an alternative strategy: Permutation Feature Importance.
```

jjerphan commented:
Instructions given in #20301 will help you to solve the conflicts caused by code style reformatting on `main`.

jjerphan commented on sklearn/ensemble/_forest.py:
```diff
@@ -174,6 +178,40 @@ def _parallel_build_trees(tree, forest, X, y, sample_weight, tree_idx, n_trees,
+    if sample_weight is None:
+        sample_weight = np.ones(n_samples)
+
+    return permutation_importance(
+        estimator,
+        X[unsampled_indices, :],
+        y[unsampled_indices],
+        scoring=None,
+        n_repeats=1,
+        n_jobs=1,
+        random_state=random_state,
+        sample_weight=sample_weight[unsampled_indices],
+    ).importances[:, 0]
```
I do not know `permutation_importance` sufficiently, but at first sight, expliciting why `n_jobs=1` is used would helps: IIUC as `_permutation_importances_oob` is called within `joblib.Parallel.__apply__` and as `permutation_importance` also calls it, we explicitly force sequential execution by setting `n_jobs=1`.

Also, why does one only takes `.importances[:, 0]`? Does `permutation_importance` compute more than what we need here?

jjerphan commented on sklearn/ensemble/_forest.py:
```diff
@@ -426,38 +494,111 @@ def fit(self, X, y, sample_weight=None):
+            self.max_samples,
+        )
+
+        with config_context(assume_finite=True):
+            # avoid redundant checking performed on X in the permutation
+            # importance function.
+            oob_importances = np.transpose(Parallel(n_jobs=self.n_jobs)(
+                delayed(_permutation_importances_oob)(
+                    estimator,
+                    X,
+                    y,
+                    sample_weight,
+                    n_samples,
+                    n_samples_bootstrap,
+                    random_state,
+                )
+                for estimator in self.estimators_
+            ))
+
+        return oob_importances
```
Should we prefer threads here?

```suggestion
        with config_context(assume_finite=True):
            # avoid redundant checking performed on X in the permutation
            # importance function.
            parallel_args = {
                **_joblib_parallel_args(prefer="threads"),
                "n_jobs": self.n_jobs
            }
            oob_importances = Parallel(**parallel_args)(
                delayed(_permutation_importances_oob)(
                    estimator,
                    X,
                    y,
                    sample_weight,
                    n_samples,
                    n_samples_bootstrap,
                    random_state,
                )
                for estimator in self.estimators_
            ))
        oob_importances = np.transpose(oob_importances)
        return oob_importances
```

glemaitre replied:
I was under the impression that I did look at it before and the threading was not the best choice (not sure that we are releasing the GIL enough).

jjerphan commented on sklearn/ensemble/_forest.py:
```diff
@@ -126,8 +126,13 @@ def _generate_sample_indices(random_state, n_samples, n_samples_bootstrap):
 def _generate_unsampled_indices(random_state, n_samples, n_samples_bootstrap):
+    """Generate the indices of the OOB sample indices for an estimator.
+
+    Instead of storing the OOB sample indices in the forest, it is more memory
+    efficient to rebuild the indices given the random state used to create the
+    bootstrap. This operation can be neglected in terms of computation time
+    compared to other processes when it is used (e.g. scoring).
```
Thanks for precising it.

jjerphan reviewed (commented):
Here is my first pass on the code, including a few more comments on the examples.

I find the two examples are a bit interlinked with a bit of redundancy (this was already the case prior to this PR).

To me, another PR could merged them and introduce another example using permutation feature importance on other algorithms than `BaseForest`'s, precising its use and value.

What do you think?

robert-robison commented on examples/ensemble/plot_forest_importances.py:
```diff
@@ -75,36 +76,47 @@
-# Please see :ref:`permutation_importance` for more details. We can now plot
-# the importance ranking.
-
+# The permutation importances is more computationally costly. Indeed, it
+# requires to fit the tree and to make additional processing: each tree will
+# be evaluated on its OOB sample as well as an OOB sample where features will
+# be permuted. This step is costly and explains the time fitting difference
+# between of the two forests.
```
Yeah I think that makes sense. Maybe it can be changed to `Elapsed time to train and compute the importances`?

robert-robison committed:
Apply suggestions from code review

robert-robison committed:
Merge commit '0e7761cdc4f244adb4803f1a97f0a9fe4b365a99' into rf_permutation_importance

robert-robison committed:
black formatted changes

robert-robison committed:
Merge remote-tracking branch 'upstream/main' into rf_permutation_importance

robert-robison committed:
remove old assert_allclose import

robert-robison commented:
I agree the examples are a little redundant and that combining them would be outside the scope of this PR. Combining them seems like a good idea to me, but I also don't know what the intended purpose of the `plot_forest_importances.py` example is.

glemaitre commented on doc/modules/ensemble.rst:
```diff
@@ -301,19 +345,24 @@ a :class:`ExtraTreesClassifier` model.
 
 .. topic:: References
-
+ .. [Strobl07] `Strobl, C., Boulesteix, AL., Zeileis, A. et al.
+ Bias in random forest variable importance measures: Illustrations,
+ sources and a solution. BMC Bioinformatics 8, 25 (2007).
+ <https://doi.org/10.1186/1471-2105-8-25>`_
+ .. [White94] `White, A.P., Liu, W.Z. Technical Note:
+ Bias in Information-Based Measures in Decision Tree Induction.
+ Machine Learning 15, 321–329 (1994).
+ <https://doi.org/10.1023/A:1022694010754>`_
```
```suggestion
  .. [Strobl07] `Strobl, C., Boulesteix, AL., Zeileis, A. et al.
     Bias in random forest variable importance measures: Illustrations,
     sources and a solution.
     BMC Bioinformatics 8, 25 (2007).
     <https://doi.org/10.1186/1471-2105-8-25>`_
  .. [White94] `White, A.P., Liu, W.Z. Technical Note:
     Bias in Information-Based Measures in Decision Tree Induction.
     Machine Learning 15, 321–329 (1994).
     <https://doi.org/10.1023/A:1022694010754>`_
```

I think that it should be the error given by sphinx.

glemaitre commented on examples/ensemble/plot_forest_importances.py:
```diff
@@ -4,42 +4,50 @@
 # %%
-# Feature importance based on mean decrease in impurity
+# Feature importance based on Mean Decrease in Impurity (MDI)
 # -----------------------------------------------------
```
This line should have as many character than the previous one.

glemaitre commented:
> To me, another PR could merged them and introduce another example using permutation feature importance on other algorithms than BaseForest's, precising its use and value.

We could keep it for another PR for sure. One issue with removing examples boils down to referencing in Google. I am not sure if we can get outdated links that will not work anymore.

robert-robison committed:
Apply suggested format changes

robert-robison committed:
Reference section edits

jjerphan commented:
> We could keep it for another PR for sure. One issue with removing examples boils down to referencing in Google. I am not sure if we can get outdated links that will not work anymore.

I agree. The content of [`plot_permutation_importance`](https://scikit-learn.org/stable/auto_examples/inspection/plot_permutation_importance.html) could complete the one on forests' feature importance, [`plot_forest_importances`](https://scikit-learn.org/stable/auto_examples/ensemble/plot_forest_importances.html).

This would allow replacing its content with a generic example for permutation feature importance using other interfaces.
This would also allow could the link as it, even though the content would be different between different versions' documentation.

What do you think?

glemaitre commented:
This sound good to me

On Wed, 23 Jun 2021 at 09:09, Julien Jerphanion ***@***.***>
wrote:

> We could keep it for another PR for sure. One issue with removing examples
> boils down to referencing in Google. I am not sure if we can get outdated
> links that will not work anymore.
>
> I agree. The content of plot_permutation_importance
> <https://scikit-learn.org/stable/auto_examples/inspection/plot_permutation_importance.html>
> could complete the one on forests' feature importance,
> plot_forest_importances
> <https://scikit-learn.org/stable/auto_examples/ensemble/plot_forest_importances.html>
> .
>
> This would allow replacing its content with a generic example for
> permutation feature importance using other interfaces.
> This would also allow could the link as it, even though the content would
> be different between different versions' documentation.
>
> What do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/18603#issuecomment-866588221>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABY32P4G7WLQ7JYYMI7Z2LTTUGCBLANCNFSM4SMAN7WA>
> .
>


-- 
Guillaume Lemaitre
Scikit-learn @ Inria Foundation
https://glemaitre.github.io/

lorentzenchr commented:
Without reading this whole PR: Note that OOB permutation importance is biased und should be corrected, see [The revival of the Gini importance](https://academic.oup.com/bioinformatics/article/34/21/3711/4994791). This corrected version (AIR) is (among other alternatives) implemented in the ranger package.

robert-robison commented:
> Without reading this whole PR: Note that OOB permutation importance is biased und should be corrected, see [The revival of the Gini importance](https://academic.oup.com/bioinformatics/article/34/21/3711/4994791). This corrected version (AIR) is (among other alternatives) implemented in the ranger package.

I think you have impurity importance and OOB permutation importance confused. I don't believe that paper found permutation importance to be biased. The following is a quote from the introduction:
> The impurity importance is known to be biased in favor of variables with many possible split points...The permutation importance does not suffer from these kinds of bias and is consequently generally preferred

Their experiments confirm this--the impurity importance is biased towards variables with more split points, while the permutation importance is unbiased. The only negative comment they make towards permutation importance is that positive outliers are more likely due to overlap in the out-of-bag observations.

AIR is not corrected OOB permutation importance, but corrected impurity importance (it stands for Actual Impurity Reduction). I would be all in favor of implementing that as well, but I think that would probably have to be in a subsequent PR, as it would require some pretty significant changes:
> As explained in Section 2.4, the tree growing procedure is slightly altered when the AIR importance is computed. Because a fraction of the splits are uninformative, this could lead to a loss of prediction accuracy of the RF, in particular if the mtry value is low. Consequently, we recommend to grow a separate RF for prediction when the AIR importance is used and the ranger package issues a warning when prediction is performed with a forest grown with the AIR importance.

Let me know if my understanding is incorrect in any way

lorentzenchr commented:
@robert-robison Thanks for your clarifications and helping my confusion. Then I'm keen to see the progress of this PR. Explainability tools are very useful.

jjerphan commented on examples/ensemble/plot_forest_importances.py:
```diff
@@ -4,23 +4,24 @@
-remaining are not.
+features on an artificial classification task.
+
+We show two strategies to estimate the feature importances: (i) the
+impurity-based feature importances and (ii) the permutation feature
+importances on out-of-bag (OOB) samples.
+.. warning::
+    Impurity-based feature importances can be misleading for high cardinality
+    features (many unique values). Check the documentation of the
+    `feature_importances` parameter to have more details regarding the
+    alternative as the permutation feature importances.
```
Small up @robert-robison.

jjerphan commented on examples/ensemble/plot_forest_importances.py:
```diff
@@ -75,36 +75,45 @@
 # %%
```
This thread can be resolved.

jjerphan reviewed (changes requested):
This looks good to me after merge conflict resolution and one remaining comment.

glemaitre commented:
@jjerphan Am I not sure if we settle on a possible API thought since we might have other metrics that would provide other feature importance: https://github.com/scikit-learn/scikit-learn/issues/20059#issuecomment-869811256

robert-robison committed:
Update examples/ensemble/plot_forest_importances.py

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

adrinjalali commented:
Removing the milestone, will have this in once the API discussion is resolved :)
